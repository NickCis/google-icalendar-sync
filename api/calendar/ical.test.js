const mockCalendar = {
  events: {
    list: jest.fn(),
  },
};

jest.mock('micro', () => ({ send: jest.fn() }));
jest.mock('micro-query', () => jest.fn());
jest.mock('micronize', () => jest.fn(f => f));
jest.mock('micro-morgan', () => jest.fn(() => f => f));
jest.mock('googleapis', () => ({
  google: {
    calendar: jest.fn().mockReturnValue(mockCalendar),
  },
}));
jest.mock('../utils/OAuth2Client', () =>
  jest.fn(() => ({
    setCredentials: jest.fn(),
  }))
);

const micro = require('micro');
const query = require('micro-query');
const ical = require('./ical');

beforeEach(() => {
  micro.send.mockReset();
  query.mockReset();
  mockCalendar.events.list.mockReset();
});

it('should send 401 if no token', async () => {
  const res = {};
  query.mockReturnValue({
    id: 'id',
  });

  await ical({}, res);

  expect(micro.send).toHaveBeenCalledWith(res, 401);
});

it('should build ical', async () => {
  const req = {
    headers: {
      host: 'test.host',
    },
  };
  const res = {
    setHeader: jest.fn(),
  };

  query.mockReturnValue({
    t: JSON.stringify('token'),
    id: 'id',
  });

  mockCalendar.events.list.mockReturnValue(
    Promise.resolve({
      config: {},
      data: {
        kind: 'calendar#events',
        etag: '"etab"',
        summary: 'Calendar title',
        updated: '2019-03-13T03:54:30.086Z',
        timeZone: 'America/Argentina/Buenos_Aires',
        accessRole: 'owner',
        defaultReminders: [
          {
            method: 'popup',
            minutes: 10,
          },
          {
            method: 'email',
            minutes: 10,
          },
        ],
        items: [
          {
            kind: 'calendar#event',
            etag: '"etag"',
            id: 'item-id',
            status: 'tentative',
            htmlLink:
              'https://www.google.com/calendar/event?eid=XXXXXXXXXXXXXX',
            created: '2012-04-09T04:03:21.000Z',
            updated: '2012-04-09T04:03:22.000Z',
            summary: 'Event title',
            description: 'Description of event',
            creator: {
              email: 'email@gmail.com',
              displayName: 'My Name',
              self: true,
            },
            organizer: {
              email: 'email@gmail.com',
              displayName: 'My Name',
              self: true,
            },
            start: {
              dateTime: '2012-05-05T09:00:00-03:00',
            },
            end: {
              dateTime: '2012-05-05T13:00:00-03:00',
            },
            iCalUID: 'item-id@google.com',
            sequence: 1,
            attendees: [
              {
                email: 'email@gmail.com',
                displayName: 'My Name',
                organizer: true,
                self: true,
                responseStatus: 'accepted',
              },
            ],
            extendedProperties: {
              shared: {
                'CalendarSyncAdapter#originalTimezone':
                  'America/Argentina/Buenos_Aires',
              },
            },
            reminders: {
              useDefault: true,
            },
          },
          {
            kind: 'calendar#event',
            etag: '"etag"',
            id: 'item-id-2',
            status: 'confirmed',
            htmlLink: 'https://www.google.com/calendar/event?eid=XXXXXXXXX',
            created: '2016-07-12T01:57:03.000Z',
            updated: '2016-07-12T01:57:03.477Z',
            summary: 'Yearly recurring event',
            creator: {
              email: 'email@gmail.com',
              displayName: 'My Name',
              self: true,
            },
            organizer: {
              email: 'email@gmail.com',
              displayName: 'My Name',
              self: true,
            },
            start: {
              date: '2016-07-11',
            },
            end: {
              date: '2016-07-12',
            },
            recurrence: ['RRULE:FREQ=YEARLY'],
            transparency: 'transparent',
            iCalUID: 'item-id-2@google.com',
            sequence: 0,
            reminders: {
              useDefault: true,
            },
          },
          {
            kind: 'calendar#event',
            etag: '"etag"',
            id: 'item-id-3',
            status: 'confirmed',
            htmlLink: 'https://www.google.com/calendar/event?eid=XXXXXXXXX',
            created: '2016-07-12T01:57:03.000Z',
            updated: '2016-07-12T01:57:03.477Z',
            summary: 'Weekly wednesday recurring event',
            creator: {
              email: 'email@gmail.com',
              displayName: 'My Name',
              self: true,
            },
            organizer: {
              email: 'email@gmail.com',
              displayName: 'My Name',
              self: true,
            },
            start: {
              date: '2016-07-11',
            },
            end: {
              date: '2016-07-12',
            },
            recurrence: ['RRULE:FREQ=WEEKLY;BYDAY=WE'],
            transparency: 'transparent',
            iCalUID: 'item-id-3@google.com',
            sequence: 0,
            reminders: {
              useDefault: true,
            },
          },
          {
            kind: 'calendar#event',
            etag: '"etag"',
            id: 'item-id-4',
            status: 'confirmed',
            htmlLink: 'https://www.google.com/calendar/event?eid=XXXXXXXXXXXX',
            created: '2019-03-16T05:55:02.000Z',
            updated: '2019-03-16T05:55:02.849Z',
            summary: 'Test: Cada mes 3er miercoles',
            creator: {
              email: 'email@gmail.com',
              displayName: 'My Name',
              self: true,
            },
            organizer: {
              email: 'email@gmail.com',
              displayName: 'My Name',
              self: true,
            },
            start: {
              date: '2019-03-20',
            },
            end: {
              date: '2019-03-21',
            },
            recurrence: ['RRULE:FREQ=MONTHLY;BYDAY=3WE'],
            // "RRULE:FREQ=MONTHLY;COUNT=3;INTERVAL=3;BYDAY=3WE"
            // "RRULE:FREQ=MONTHLY;UNTIL=20190620;INTERVAL=3"
            transparency: 'transparent',
            iCalUID: 'item-id-4@google.com',
            sequence: 0,
            extendedProperties: {
              private: {
                everyoneDeclinedDismissed: '-1',
              },
            },
            reminders: {
              useDefault: true,
            },
          },
          {
            kind: 'calendar#event',
            etag: '"etag"',
            id: 'item-id-5',
            status: 'cancelled',
            recurringEventId: 'item-id-4',
            originalStartTime: {
              date: '2019-04-17',
            },
          },
          {
            kind: 'calendar#event',
            etag: '"etag"',
            id: 'item-id-6',
            status: 'cancelled',
            recurringEventId: 'item-id-7',
            originalStartTime: {
              dateTime: '2019-03-16T10:00:00.000',
              timeZone: 'America/Buenos_Aires',
            },
          },
          {
            kind: 'calendar#event',
            etag: '"etag"',
            id: 'item-id-7',
            status: 'confirmed',
            htmlLink: 'https://www.google.com/calendar/event?eid=XXXXXXXXXXXX',
            created: '2019-03-16T05:55:02.000Z',
            updated: '2019-03-16T05:55:02.849Z',
            summary: 'Test: Cada mes 3er miercoles',
            creator: {
              email: 'email@gmail.com',
              displayName: 'My Name',
              self: true,
            },
            organizer: {
              email: 'email@gmail.com',
              displayName: 'My Name',
              self: true,
            },
            start: {
              dateTime: '2019-03-16T10:00:00.000',
              timeZone: 'America/Buenos_Aires',
            },
            end: {
              dateTime: '2019-03-16T11:00:00.000',
              timeZone: 'America/Buenos_Aires',
            },
            recurrence: ['RRULE:FREQ=MONTHLY;BYDAY=3WE'],
            transparency: 'transparent',
            iCalUID: 'item-id-7@google.com',
            sequence: 0,
            extendedProperties: {
              private: {
                everyoneDeclinedDismissed: '-1',
              },
            },
            reminders: {
              useDefault: true,
            },
          },
        ],
      },
    })
  );

  await ical(req, res);

  expect(res.setHeader).toHaveBeenCalledWith(
    'Content-Type',
    'text/calendar; charset=UTF-8'
  );

  expect(
    micro.send.mock.calls[0][2].replace(/^DTSTAMP:.+$/gm, 'DTSTAMP:')
  ).toMatchSnapshot();
});
