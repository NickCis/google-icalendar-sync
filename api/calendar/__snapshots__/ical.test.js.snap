// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should build ical 1`] = `
"BEGIN:VCALENDAR
VERSION:2.0
PRODID:-//Open Source Software//google-icalendar-sync//EN
NAME:Calendar title
X-WR-CALNAME:Calendar title
TIMEZONE-ID:America/Argentina/Buenos_Aires
X-WR-TIMEZONE:America/Argentina/Buenos_Aires
BEGIN:VEVENT
UID:item-id@test.host
SEQUENCE:1
DTSTAMP:
DTSTART:20120505T120000Z
DTEND:20120505T160000Z
SUMMARY:Event title
DESCRIPTION:Description of event
ORGANIZER;CN=\\"My Name\\":mailto:email@gmail.com
ATTENDEE;ROLE=REQ-PARTICIPANT;CN=\\"My Name\\":MAILTO:email@gmail.com
URL;VALUE=URI:https://www.google.com/calendar/event?eid=XXXXXXXXXXXXXX
STATUS:TENTATIVE
CREATED:20120409T040321Z
LAST-MODIFIED:20120409T040322Z
END:VEVENT
BEGIN:VEVENT
UID:item-id-2@test.host
SEQUENCE:0
DTSTAMP:
DTSTART;VALUE=DATE:20160711
DTEND;VALUE=DATE:20160712
X-MICROSOFT-CDO-ALLDAYEVENT:TRUE
X-MICROSOFT-MSNCALENDAR-ALLDAYEVENT:TRUE
RRULE:FREQ=YEARLY
SUMMARY:Yearly recurring event
ORGANIZER;CN=\\"My Name\\":mailto:email@gmail.com
URL;VALUE=URI:https://www.google.com/calendar/event?eid=XXXXXXXXX
STATUS:CONFIRMED
CREATED:20160712T015703Z
LAST-MODIFIED:20160712T015703Z
END:VEVENT
BEGIN:VEVENT
UID:item-id-3@test.host
SEQUENCE:0
DTSTAMP:
DTSTART;VALUE=DATE:20160711
DTEND;VALUE=DATE:20160712
X-MICROSOFT-CDO-ALLDAYEVENT:TRUE
X-MICROSOFT-MSNCALENDAR-ALLDAYEVENT:TRUE
RRULE:FREQ=WEEKLY;BYDAY=WE
SUMMARY:Weekly wednesday recurring event
ORGANIZER;CN=\\"My Name\\":mailto:email@gmail.com
URL;VALUE=URI:https://www.google.com/calendar/event?eid=XXXXXXXXX
STATUS:CONFIRMED
CREATED:20160712T015703Z
LAST-MODIFIED:20160712T015703Z
END:VEVENT
BEGIN:VEVENT
UID:item-id-4@test.host
SEQUENCE:0
DTSTAMP:
DTSTART;VALUE=DATE:20190320
DTEND;VALUE=DATE:20190321
X-MICROSOFT-CDO-ALLDAYEVENT:TRUE
X-MICROSOFT-MSNCALENDAR-ALLDAYEVENT:TRUE
RRULE:FREQ=MONTHLY;BYDAY=3WE
EXDATE;VALUE=DATE:20190417
SUMMARY:Test: Cada mes 3er miercoles
ORGANIZER;CN=\\"My Name\\":mailto:email@gmail.com
URL;VALUE=URI:https://www.google.com/calendar/event?eid=XXXXXXXXXXXX
STATUS:CONFIRMED
CREATED:20190316T055502Z
LAST-MODIFIED:20190316T055502Z
END:VEVENT
BEGIN:VEVENT
UID:item-id-7@test.host
SEQUENCE:0
DTSTAMP:
DTSTART;TZID=America/Buenos_Aires:20190316T100000
DTEND;TZID=America/Buenos_Aires:20190316T110000
RRULE:FREQ=MONTHLY;BYDAY=3WE
EXDATE:20190316T130000Z
SUMMARY:Test: Cada mes 3er miercoles
ORGANIZER;CN=\\"My Name\\":mailto:email@gmail.com
URL;VALUE=URI:https://www.google.com/calendar/event?eid=XXXXXXXXXXXX
STATUS:CONFIRMED
CREATED:20190316T055502Z
LAST-MODIFIED:20190316T055502Z
END:VEVENT
END:VCALENDAR"
`;

exports[`should handle recurring events moved 1`] = `
"BEGIN:VCALENDAR
VERSION:2.0
PRODID:-//Open Source Software//google-icalendar-sync//EN
NAME:Calendar title
X-WR-CALNAME:Calendar title
TIMEZONE-ID:America/Argentina/Buenos_Aires
X-WR-TIMEZONE:America/Argentina/Buenos_Aires
BEGIN:VEVENT
UID:recurring-id_R20180822T141500@test.host
SEQUENCE:2
DTSTAMP:
DTSTART:20190124T151500Z
DTEND:20190124T153000Z
RECURRENCE-ID:20190124T141500Z
SUMMARY:My title
DESCRIPTION:Some description
ORGANIZER;CN=\\"Another One Mail\\":mailto:another-one@mail.com
ATTENDEE;ROLE=REQ-PARTICIPANT;CN=\\"Person Mail\\":MAILTO:Person@mail.com
ATTENDEE;ROLE=REQ-PARTICIPANT;CN=\\"Me Self\\":MAILTO:me@mail.com
STATUS:CONFIRMED
CREATED:20180702T141020Z
LAST-MODIFIED:20190322T140618Z
END:VEVENT
END:VCALENDAR"
`;
